### PR TITLE
add CMakeList.txt. build windows binary by using vs2019.

### PR DESCRIPTION
**Pull request by chenhuitao**
Only tested in vs2019.
Open the "x86 Native Tools Command Prompt for VS 2019",
  cd ${UNICORN_SOURCE_DIR}
  mkdir build
  cd build
  cmake -G "NMake Makefiles" ..
  nmake
or Open "x64 Native Tools Command Prompt for VS 2019" for 64bit binary.
Build unicorn.dll and samples.

### CODE REVIEW

**Comment by Waterman178** [comment]
mark

**Comment by aquynh** [comment]
excellent work! 

i havent tried with MSVC, but i did this on Ubuntu x64:

```
mkdir build; cd build
cmake ..; make
```

this is a long process, but eventually it builds libunicorn.so, and i got below errors:

```
...
[ 94%] Linking C shared library libunicorn.so
/usr/bin/ld: libarm-softmmu.a(tcg.c.o): relocation R_X86_64_PC32 against symbol `stderr@@GLIBC_2.2.5' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
CMakeFiles/unicorn.dir/build.make:1120: recipe for target 'libunicorn.so' failed
make[2]: *** [libunicorn.so] Error 1
CMakeFiles/Makefile2:716: recipe for target 'CMakeFiles/unicorn.dir/all' failed
make[1]: *** [CMakeFiles/unicorn.dir/all] Error 2
Makefile:83: recipe for target 'all' failed
make: *** [all] Error 2
```

also, this builds bunch of .a files, such as `libmips64-softmmu.a`. is this necessary?

**Comment by aquynh** [comment]
please just commit your next changes to this same fork, so later we will just merge them all

**Comment by chenhuitao** [comment]
I will improve it support linux, windows and mac.

**Comment by aquynh** [comment]
i tried to build on Windows, and got some warnings like below:

```
[ 92%] Building C object CMakeFiles/unicorn.dir/msvc/unicorn/qapi-types.c.obj
qapi-types.c
C:\Users\me\Documents\unicorn-cmake\include\unicorn/platform.h(208) : warning C4005: 'va_copy' : macro redefinition
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\stdarg.h(37) : see previous definition of 'va_copy'
[ 93%] Building C object CMakeFiles/unicorn.dir/msvc/unicorn/qapi-visit.c.obj
qapi-visit.c
C:\Users\me\Documents\unicorn-cmake\include\unicorn/platform.h(208) : warning C4005: 'va_copy' : macro redefinition
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\stdarg.h(37) : see previous definition of 'va_copy'
[ 93%] Linking C shared library unicorn.dll
   Creating library unicorn.lib and object unicorn.exp
[ 93%] Built target unicorn
Scanning dependencies of target sample_arm64eb
[ 94%] Building C object CMakeFiles/sample_arm64eb.dir/samples/sample_arm64eb.c.obj
sample_arm64eb.c
c:\users\me\documents\unicorn-cmake\include\unicorn\platform.h(164) : warning C4996: '_ftime64': This function or variable may be unsafe. Consider using _ftime64_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS. See online help for details.
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\sys/timeb.h(131) : see declaration of '_ftime64'
c:\users\me\documents\unicorn-cmake\include\unicorn\platform.h(208) : warning C4005: 'va_copy' : macro redefinition
        C:\Program Files (x86)\Microsoft Visual Studio 12.0\VC\INCLUDE\stdarg.h(37) : see previous definition of 'va_copy'
[ 94%] Linking C executable sample_arm64eb.exe
```

please fix the warnings.

otherwise, all sample*.exe seem to work, great job!

**Comment by aquynh** [comment]
FYI, I am using VS2013.

**Comment by aquynh** [comment]
we need a friendly script to build with cmake. please adapt https://github.com/aquynh/capstone/blob/next/nmake.bat for that.

**Comment by chenhuitao** [comment]
Has been tested on windows and linux.
nmake.bat [x86] [arm] [aarch64] [m68k] [mips] [sparc]
cmake [-DUNICORN_ARCH="[x86] [arm] [aarch64] [m68k] [mips] [sparc]"]
Build all by default.

File: nmake.bat [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+shift /0
+goto loop
+
+:end
+if "%allparams%"=="" (
+    goto eof
+)
+:: remove left, right blank
+:intercept_left
+if "%allparams:~0,1%"==" " set "allparams=%allparams:~1%"&goto intercept_left
```
Comment (aquynh) [inline nmake.bat]: do we need a space after "&"?

(in 2 lines below too)

**Comment by aquynh** [comment]
trying to build on Windows with MSVC 2013, i got the below errors.

```
...
Scanning dependencies of target arm-softmmu
[  1%] Building C object CMakeFiles/arm-softmmu.dir/qemu/cpu-exec.c.obj
cpu-exec.c
[  1%] Building C object CMakeFiles/arm-softmmu.dir/qemu/cpus.c.obj
cpus.c
[  1%] Building C object CMakeFiles/arm-softmmu.dir/qemu/cputlb.c.obj
cputlb.c
[  1%] Building C object CMakeFiles/arm-softmmu.dir/qemu/exec.c.obj
exec.c
[  2%] Building C object CMakeFiles/arm-softmmu.dir/qemu/fpu/softfloat.c.obj
softfloat.c
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(188) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(189) : error C2085: 'set_float_detect_tininess_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(189) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(192) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(193) : error C2085: 'set_float_rounding_mode_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(193) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(196) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(197) : error C2085: 'set_float_exception_flags_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(197) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(200) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(201) : error C2085: 'set_floatx80_rounding_precision' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(201) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(204) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(205) : error C2085: 'set_flush_to_zero_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(205) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(208) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(209) : error C2085: 'set_flush_inputs_to_zero_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(209) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(212) : error C2054: expected '(' to follow 'inline'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(213) : error C2085: 'set_default_nan_mode_arm' : not in formal parameter list
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(213) : error C2143: syntax error : missing ';' before '{'
C:\Users\me\Documents\unicorn-cmake\qemu\include\fpu/softfloat.h(216) : error C2054: expected '(' to follow 'inline'
C
....
```

do you build without any issues on MSVC 2019?

**Comment by aquynh** [comment]
on WSL, built with cmake, i got this error when running sample_mips.

```
$  ./sample_mips
Emulate MIPS code (big-endian)
>>> Tracing basic block at 0x10000, block size = 0x4
>>> Tracing instruction at 0x10000, instruction size = 0x4
>>> Emulation done. Below is the CPU context
>>> R1 = 0x77df
===========================
Emulate MIPS code (little-endian)
Failed on uc_open() with error returned: 2 (Invalid/unsupported architecture (UC_ERR_ARCH))
```

**Comment by aquynh** [comment]
meanwhile, on `master` we can run `sample_mips` without any issues

```
$ ./samples/sample_mips
Emulate MIPS code (big-endian)
>>> Tracing basic block at 0x10000, block size = 0x4
>>> Tracing instruction at 0x10000, instruction size = 0x4
>>> Emulation done. Below is the CPU context
>>> R1 = 0x77df
===========================
Emulate MIPS code (little-endian)
>>> Tracing basic block at 0x10000, block size = 0x4
>>> Tracing instruction at 0x10000, instruction size = 0x4
>>> Emulation done. Below is the CPU context
>>> R1 = 0x77df

```

File: nmake.bat (hunk 1 above)
Comment (chenhuitao) [reply nmake.bat]: Yes, you are right.

**Comment by chenhuitao** [comment]
1. VS2019 does not has this problem. VS2013 does not support 'inline', must be '__inline', added a macro.
2. Macro definition of mipsel is wrong, has been fixed in last commit.

**Comment by chenhuitao** [comment]
Installation not support windows.
Default CMAKE_INSTALL_PREFIX on linux is '/usr/local'.
Add the cmake.sh.

**Comment by aquynh** [comment]
the COMPILE_CMAKE.TXT still has some architectures existing in Capstone, but not in Unicorn (like EVM, etc). please read-proof it

**Comment by aquynh** [comment]
and please put a link to COMPILE_CMAKE in https://github.com/unicorn-engine/unicorn/blob/master/docs/COMPILE.md

**Comment by chenhuitao** [comment]
Fixed.

**Comment by aquynh** [comment]
merged, thanks for this amazing work!

@stephengroat, can you please integrate cmake with Travis & Appveyor now?

**Pull request merged**